### ISSUE TITLE
Harbor handles "307 Temporary Redirect"  incorrectly

### ISSUE THREAD

**Issue by kdada** [vmware/harbor#1067]
Version:
- docker 1.12.0    
- docker-compose 1.8.1
- harbor 0.4.0  

Registry use aliyun OSS as storage backend.   

When I synced a harbor to another,harbor gave some errors:
```
2016-11-11T07:14:57Z [ERROR] [transfer.go:379]: an error occurred while pulling blob sha256:56bec22e355981d8ba0878c6c2f23b21f422f30ab0aba188b54f1ffeff59c190 of 000uq/test-deploy-service:v0.1.0-deploy from http://registry:5000: 400 <?xml version="1.0" encoding="UTF-8"?>
<Error>
  <Code>InvalidArgument</Code>
  <Message>Either the Signature query string parameter or the Authorization header should be specified, not both.</Message>
  <RequestId>58256FF1B47FB80CA4F629F3</RequestId>
  <HostId>caicloud-registry.oss-cn-hangzhou.aliyuncs.com</HostId>
  <Authorization>Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJSUzI1NiIsImtpZCI6IkRTUks6QTMyVjpLTFZSOlBNRlg6SkpVQzpTWTZaOjI2M1Y6SEJOUzo2M0U3OjJVTFA6TkNQSDpGTDRGIn0.eyJpc3MiOiJyZWdpc3RyeS10b2tlbi1pc3N1ZXIiLCJzdWIiOiJqb2Itc2VydmljZS11c2VyIiwiYXVkIjoidG9rZW4tc2VydmljZSIsImV4cCI6MTQ3ODg1MDI5NSwibmJmIjoxNDc4ODQ4NDk1LCJpYXQiOjE0Nzg4NDg0OTUsImp0aSI6InEybGFoeDhpSlhVM1ZQN3ciLCJhY2Nlc3MiOlt7InR5cGUiOiJyZXBvc2l0b3J5IiwibmFtZSI6IjAwMHVxL3Rlc3QtZGVwbG95LXNlcnZpY2UiLCJhY3Rpb25zIjpbInB1bGwiLCJwdXNoIiwiKiJdfV19.Rn-s8txunMqAV8M0Ea-NvkGnpxUv5uXmn-yeZr9TQ9CzOAViR2XhX_sip3XghdP9stEF1JXLYzitdd6kPNAqjqaDo6VDO7OtGdfF_UbolbXp0p20Sn7iBgJrcfHDKWRtxFLE231bhqlxffs11cbb5qCsVV8CM1zP0lkSTkEcvp1L7tkELDBbee5GYOl141gKQLuH9BFy3Z4cUQhRqe-mjQ2yVJRWMaaocrxnlRH2Y0utqkAQZiC8Bcq5jpf8MBtUQSP26TbwpaUqTEfQtKByVZYe08bvBpYp3fdTRZueHvQeolgXsVbwSJ0ecaaBCL49bmLvgXJaS1qdD7lUyerBjhoyJgMCvEqtTcJxmMsnaQvQnco3WErWp2Ov0gJTfV7e-aHTdANlRAg-Sx08cyR8UDD6UJnzalPXFzujLGKbjl6T4s7q9XmlRFP49FVm0Hb_iL9Re-QdgYquhj9XgMBnBJpAWQZyf7i4qT5DRjo2nqTfvysVVbj-uJOXVwalwjgchSMKk3CwzgttlHtZM7___rU5UcACPwxHnQZ138q3_TrSak3HYgprAV4KQ78P6bGLXY5bblB9zONDR9uHm0RE4Gy0nGkvcjN9srRSSRUX0DpLv5lc2JSvENfzpJJhCsiD1eNZmESSDsFKGYXPdC4UfFkPKHZFaOy76HZAMj0oCnA</Authorization>
</Error>
```

Response of Registry:
```
< HTTP/1.1 307 Temporary Redirect
< Content-Type: application/octet-stream
< Docker-Distribution-Api-Version: registry/2.0
< Location: https://caicloud-registry.oss-cn-hangzhou.aliyuncs.com/docker/registry/v2/blobs/sha256/56/56bec22e355981d8ba0878c6c2f23b21f422f30ab0aba188b54f1ffeff59c190/data?Expires=1478852247&OSSAccessKeyId=*******************&Signature=*******************
< Date: Fri, 11 Nov 2016 07:57:27 GMT
< Content-Length: 297
< 
<a href="https://caicloud-registry.oss-cn-hangzhou.aliyuncs.com/docker/registry/v2/blobs/sha256/56/56bec22e355981d8ba0878c6c2f23b21f422f30ab0aba188b54f1ffeff59c190/data?Expires=1478852247&amp;OSSAccessKeyId=*******************&amp;Signature=*******************">Temporary Redirect</a>.
```

OSS can't handle a request which has "Signature" and "Authorization" at the same time.  
It needs to delete "Authorization" from request header when harbor got a "Temporary Redirect".

**Comment by reasonerjt** [comment]
This should reflect the same root cause of #1006

**Comment by reasonerjt** [comment]
Adding the redirect interceptor to remove the header should fix such issue.

**Comment by kdada** [comment]
@reasonerjt 

It does not work.  

Harbor use "utils/registry.Transport" to modify requests(Add Auth).  
In http.Client, CheckRedirect is called before Transport.RoundTrip().  
So I can't remove header in CheckRedirect.  

I solve it by adding a "HostRoundTrip":
```go
// HostRoundTrip bind different RoundTripper by host
type HostRoundTrip struct {
	host    string
	hostRT  http.RoundTripper
	otherRT http.RoundTripper
}

// NewHostRoundTrip create HostRoundTrip
func NewHostRoundTrip(endpoint string, hostRT http.RoundTripper, otherRT http.RoundTripper) (*HostRoundTrip, error) {
	u, err := url.Parse(endpoint)
	return &HostRoundTrip{
		u.Host,
		hostRT,
		otherRT,
	}, err
}

// RoundTrip check host of req
func (rt *HostRoundTrip) RoundTrip(req *http.Request) (*http.Response, error) {
	if rt.host == req.Host {
		log.Debug(req.URL.String(), " use host RoundTrip")
		return rt.hostRT.RoundTrip(req)
	}
	log.Debug(req.URL.String(), " use other RoundTrip")
	return rt.otherRT.RoundTrip(req)
}



// NewRepository returns an instance of Repository
func NewRepository(name, endpoint string, client *http.Client) (*Repository, error) {
	name = strings.TrimSpace(name)

	u, err := utils.ParseEndpoint(endpoint)
	if err != nil {
		return nil, err
	}
        // ADD:
	client.Transport, err = NewHostRoundTrip(endpoint, client.Transport, http.DefaultTransport)
	if err != nil {
		return nil, err
	}

	repository := &Repository{
		Name:     name,
		Endpoint: u,
		client:   client,
	}

	return repository, nil
}
```

**Comment by ywk253100** [comment]
@kdada Thank you for reporting this.
@reasonerjt We can do some filter to the requests, if the request is sent to registry(e.g. contains "v2"), we add authentication header, otherwise we don't.

**Comment by ywk253100** [comment]
Fixed by #1224 .
